src/sshagent/OpenSSHKey.cpp

```diff
@@ -0,0 +1,480 @@
+    , m_cipherName(other.m_cipherName)
+    , m_kdfName(other.m_kdfName)
+    , m_kdfOptions(other.m_kdfOptions)
+    , m_rawPrivateData(other.m_rawPrivateData)
+    , m_publicData(other.m_publicData)
+    , m_privateData(other.m_privateData)
+    , m_comment(other.m_comment)
+    , m_error(other.m_error)
+{
+
+}
+
+bool OpenSSHKey::operator==(const OpenSSHKey& other) const
+{
+    // close enough for now
+    return (fingerprint() == other.fingerprint());
+}
+
+const QString OpenSSHKey::cipherName() const
+{
+    return m_cipherName;
+}
+
+const QString OpenSSHKey::type() const
+{
+    return m_type;
+}
+
+int OpenSSHKey::keyLength() const
+{
+    if (m_type == "ssh-dss" && m_publicData.length() == 4) {
+        return (m_publicData[0].length() - 1) * 8;
+    } else if (m_type == "ssh-rsa" && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 8;
+    } else if (m_type.startsWith("ecdsa-sha2-") && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 4;
+    } else if (m_type == "ssh-ed25519" && m_publicData.length() == 1) {
+        return m_publicData[0].length() * 8;
+    }
+
+    return 0;
+}
+
+const QString OpenSSHKey::fingerprint() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    QByteArray rawHash = QCryptographicHash::hash(publicKey, QCryptographicHash::Sha256);
+
+    return "SHA256:" + QString::fromLatin1(rawHash.toBase64(QByteArray::OmitTrailingEquals));
+}
+
+const QString OpenSSHKey::comment() const
+{
+    return m_comment;
+}
+
+const QString OpenSSHKey::publicKey() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    return m_type + " " + QString::fromLatin1(publicKey.toBase64()) + " " + m_comment;
+}
+
+const QString OpenSSHKey::errorString() const
+{
+    return m_error;
+}
+
+void OpenSSHKey::setType(const QString& type)
+{
+    m_type = type;
+}
+
+void OpenSSHKey::setPublicData(const QList<QByteArray>& data)
+{
+    m_publicData = data;
+}
+
+void OpenSSHKey::setPrivateData(const QList<QByteArray>& data)
+{
+    m_privateData = data;
+}
+
+void OpenSSHKey::setComment(const QString& comment)
+{
+    m_comment = comment;
+}
+
+void OpenSSHKey::clearPrivate()
+{
+    m_rawPrivateData.clear();
+    m_privateData.clear();
+}
+
+bool OpenSSHKey::parsePEM(const QByteArray& in, QByteArray& out)
+{
+    QString pem = QString::fromLatin1(in);
+    QStringList rows = pem.split(QRegExp("[\r\n]"), QString::SkipEmptyParts);
+
+    if (rows.length() < 3) {
+        m_error = tr("Invalid key file, expecting an OpenSSH key");
+        return false;
+    }
+
+    QString begin = rows.first();
+    QString end = rows.last();
+
+    QRegExp beginEx("-----BEGIN (.+)-----");
```

Comment:
More thorough review pending, but I just saw this while reading my notifications. This should be

`"^-----BEGIN ([^\\-]+)-----$"`

Also use QRegularExpression instead of QRegExp. Or maybe even split after -----BEGIN and before ----- using QString::indexOf().

Reply:
It's using `exactMatch` which implies `^...$`.

Reply:
Still leaves the thing in the middle.

Reply:
I had it like that first but opted to use the shorter version as it doesn't matter much with a single line exact match. Sure, it bleeds extra dashes into the capture. I'll change it back.